src/controllers/scripting/legacy/controllerscriptinterfacelegacy.h

```diff
@@ -17,6 +17,17 @@ class ConfigKey;
 class ControllerScriptInterfaceLegacy : public QObject {
     Q_OBJECT
   public:
+    enum WellKnownCharsets {
+        LATIN_9,
+        ISO_8859_15,
+        LATIN_1,
+        ISO_8859_1,
+        UCS_2,
+        ISO_10646_UCS_2
+
+    };
+    Q_ENUM(WellKnownCharsets)
```

Comment:
Dammit! I can't find how to expose this enum under `engine.WellKnownCharsets`. Can someone help me?

Reply:
You need to tell the JavaScript-Interpreter (QJSEngine) about it. Something like this:
```C++
  void registerEnums(QJSEngine& engine) {
      qRegisterMetaType<mixxx::preferences::constants::WellKnownCharsets>("WellKnownCharsets");

      const QMetaObject& metaObject = mixxx::preferences::constants::staticMetaObject;
      QMetaEnum metaEnum = metaObject.enumerator(metaObject.indexOfEnumerator("WellKnownCharsets"));

      QJSValue enumObject = engine.newObject();
      for (int enumIdx = 0; enumIdx  < metaEnum.keyCount(); ++enumIdx ) {
          enumObject.setProperty(metaEnum.key(enumIdx ), metaEnum.value(i));
      }

      engine.globalObject().setProperty("WellKnownCharsets", enumObject);
  }
```

Reply:
use `Q_ENUM`, its way simpler. https://stackoverflow.com/questions/59728876/how-to-register-an-enum-type-to-qjsengine-to-be-used-from-the-scripting-environm

Reply:
I tried that. It exposes an empty object. Here is the code I tried:

```diff
diff --git a/src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp b/src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp
index 22978c5f5c..89d9f02538 100644
--- a/src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp
+++ b/src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp
@@ -415,8 +415,11 @@ bool ControllerScriptEngineLegacy::initialize() {
     ControllerScriptInterfaceLegacy* legacyScriptInterface =
             new ControllerScriptInterfaceLegacy(this, m_logger);
 
-    engineGlobalObject.setProperty(
-            "engine", m_pJSEngine->newQObject(legacyScriptInterface));
+    auto engine = m_pJSEngine->newQObject(legacyScriptInterface);
+    auto meta = m_pJSEngine->newQMetaObject(&ControllerScriptInterfaceLegacy::staticMetaObject);
+
+    engine.setProperty("WellKnownCharsets", meta);
+    engineGlobalObject.setProperty("engine", engine);
 
 #ifdef MIXXX_USE_QML
     if (m_bQmlMode) {
```

```js
console.error(Object.keys(engine.WellKnownCharsets));
```

displays `[]`.

Reply:
Okay, shooting a little in the dark here, but have you just tried `WellKnownCharsets.LATIN_9` in the JS? is it still `undefined`?

Apart from that, try deleting the `engine.setProperty("WellKnownCharsets", meta);` line.

Reply:
Oh, no, wait, that code actually works! I'm so sorry for the noise…